clamp with max and min as tensors

vsimkus opened the pull request:
# clamp with max and min arguments as tensors

Extending the current `clamp()` functionality to take `max` and `min` arguments as either scalars or tensors. Resolves https://github.com/pytorch/pytorch/issues/2793 and makes the function more consistent with `numpy.clip`.

The `max` and `min` arguments should specify the maximum and minimum values for each element of the input tensor. The arguments can also be [broadcastable](https://pytorch.org/docs/stable/notes/broadcasting.html) tensors to the size of the main input tensor. It also allows mixing-and-matching scalar and tensor arguments, e.g. `type(min)` is a number and `type(max)` is a tensor.

Also fixes discrepancy between AVX and base vector implementation for clamp - now all implementations are consistent and use `min(max_vec, max(min_vec, x)` formula, thus making it equivalent to `numpy.clip` in all implementations.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 28caddb7:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=28caddb7e15588f02509bcbb310bc286f83291bd).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4333739) pytorch_windows_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90423483) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90423483))
<details>
<summary>
<code>CMake Error at torch_cuda_generated_UnaryOpsKernel.cu.obj.Release.cmake:221 (message):</code>
</summary>

```
Generated C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_SparseMM.cu.obj successfully.
 -D generated_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_UnaryOpsKernel.cu.obj.Release.cmake" 
FAILED: caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_UnaryOpsKernel.cu.obj  
 -D generated_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_UnaryOpsKernel.cu.obj.Release.cmake" 
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj
C:/Jenkins/Miniconda3/Library/bin/cmake.exe -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_UnaryOpsKernel.cu.obj.NVCC-depend
C:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
UnaryOpsKernel.cu 
C:/Users/circleci/project/aten/src/ATen/native/cuda/UnaryOpsKernel.cu(527): fatal error C1057: unexpected end of file in macro expansion 
CMake Error at torch_cuda_generated_UnaryOpsKernel.cu.obj.Release.cmake:221 (message):
  Error generating
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_UnaryOpsKernel.cu.obj


RING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorTransformations.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorTransformations.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorTransformations.cu.obj.Release.cmake" 
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorTransformations.cu.obj
C:/Jenkins/Miniconda3/Library/bin/cmake.exe -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/./torch_cuda_generated_TensorTransformations.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cuda/torch_cuda_generated_TensorTransformations.cu.obj.NVCC-depend
C:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
TensorTransformations.cu 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

gchanan commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -523,6 +523,70 @@ void lgamma_kernel_cuda(TensorIterator& iter) {
 }
 
+void clamp_kernel_cuda(TensorIterator& iter, Scalar min_scalar, Scalar max_scalar) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "clamp_cuda", [&]() {
```
why did you add complex?  That's not in the TH implementation, right?  And I didn't see it mentioned in the commit message.

vsimkus replied:
To my understanding the TH kernel wasn't defined for complex types. I've added complex types in the Aten cuda kernel implementation to match the Aten cpu kernel, which is defined for complex types too (support was added in https://github.com/pytorch/pytorch/commit/7c472ec5974cc638cb697b4523d88324aee38fa9). Would it be preferred to add the support for complex types for the cuda kernel in a separate PR? 

I will split the addition of the cuda kernels into a separate commit and mention the addition of complex types.

vsimkus replied:
The complex behaviour in this kernel is equivalent to that in the CPU kernel. Following the current discussion on complex behaviour, should I keep it so that the CPU and CUDA kernels are aligned, or should I remove the complex dispatcher?

Keeping in mind that complex tensors are disabled in `UnaryOps.cpp`, keeping this implementation should be ok?
https://github.com/pytorch/pytorch/blob/5d9b4d5720eff32a6997a66402b6423d703313f6/aten/src/ATen/native/UnaryOps.cpp#L264-L265

neginraoof commented:
@vsimkus Thanks for the PR! We need this PR for exporting RCNN models.

mruberry commented on test/test_torch.py:
```diff
@@ -6383,6 +6384,119 @@ def test_clamp(self, device):
         with self.assertRaisesRegex(RuntimeError, error_msg):
             m1.clamp_()
 
+    def test_clamp_with_tensors(self, device):
```
Always great to see more tests!

Before this PR is landed, contact @ailzhang to fix the PyTorch/XLA build. Otherwise landing this test will break it (as it happening in the current CI). 

Additionally, this test is really big! While a lot of the tests in test_torch.py are huge and test multiple things, we'd like to cut them up into smaller, more self-contained tests. It looks like most of these blocks are separable, and you can write a helper function to generate m1 if you like. 

Is there any place you're testing backwards?

vsimkus replied:
Thanks, I have split up the new tests for clamp_with_tensors and the old ones for clamp into smaller units.

> Is there any place you're testing backwards?
The backwards behaviour is tested by test_autograd. The test definitions for the new functionality are in `common_method_invocations.py`. It tests the backward behaviour by comparing the output from the code vs the numerical approximation using small finite distances. It also compares the outputs from the CPU kernel vs CUDA.

I will reach out to with the contact to fix the XLA build.

mruberry replied:
Just @ailzhang is sufficient. We also have a decorator, @onlyOnCPUAndCUDA that will prevent the tests from running on XLA. The XLA team can then update their behavior and remove the decorator on their own timeline.

mruberry commented on test/test_torch.py:
```diff
@@ -6383,6 +6384,119 @@ def test_clamp(self, device):
+        self.assertEqual(res1, res2)
+
+        torch.clamp(m1, max=max_vals, out=out)
+        if (out - res1).abs().max() > 0.01:
```
Is this debug code? That shouldn't be included in the PR.

vsimkus replied:
Yes, that was debug code, since I used the CI for debugging CUDA. It is now removed.

mruberry commented on torch/functional.py:
```diff
@@ -910,5 +912,112 @@ def lu(A, pivot=True, get_infos=False, out=None):
 
 
+def clamp(input, min=None, max=None, out=None):
+    r"""
```
This doc string looks pretty good, but do you need more detail about clamp's behavior on quantized and complex values? In particular, does clamping a complex value work like clamping it in Numpy? If not we should be specific about the behavior so the user knows what to expect. cc @anjali411, who's looking at complex-behavior in PyTorch.

vsimkus replied:
I have merely copied the existing docstring and added examples with tensor arguments, the description is more or less the same. 

With regards to quantized tensors, I think everything is already explained very well in the main document about [quantization](https://pytorch.org/docs/stable/quantization.html). In particular this paragraph:

>PyTorch also supports quantization aware training, which models quantization errors in both the forward and backward passes using fake-quantization modules. Note that the entire computation is carried out in floating point. At the end of quantization aware training, PyTorch provides conversion functions to convert the trained model into lower precision.

Would it then be useful to repeat it here too? I would argue that if all quantized operations behave in the same way, then it is not necessarily important to repeat it in each operation's docstring. I am happy to add it if you think it would be helpful.

With regards to complex values, since it is now a disabled feature following the recent discussions, I am not adding it to the doc.

mruberry replied:
That sounds good. What's the reason for adding clamp here, as opposed to keeping its documentation where it was?

vsimkus replied:
I've added an override of the default `torch.clamp` here to dispatch to `at::clamp_with_tensors` or `at::clamp` depending on the argument types. The build required me to move the docstring here too as otherwise I got some complaint. It is also equivalent to the other functions that are overridden/defined in `functional.py`, where their docstrings are defined within the function and not in a separate file.

mruberry replied:
Interesting. That's cool.

vsimkus commented:
@mruberry Thanks for the comments! And sorry for taking a long time to reply. 

There are a couple of debug statements in the current PR, which I will remove - I needed them to debug XLA/GPU issues, since I can't debug them on my own machine. So I would appreciate help with debugging XLA, GPU and maybe ONNX issues.

I will clean up the PR this weekend, when I've got some spare time. 

Also, I have noticed an inconsistency in the existing `clamp` kernels for different backends. I.e. the base cpu implementation of `vec256::clamp` uses a ternary operator:
https://github.com/pytorch/pytorch/blob/758ad516f32708de243f194144ee0f7b9e0f5117/aten/src/ATen/cpu/vec256/vec256_base.h#L535-L544
Whereas the AVX implementation uses nested `min` and `max` operators:
https://github.com/pytorch/pytorch/blob/758ad516f32708de243f194144ee0f7b9e0f5117/aten/src/ATen/cpu/vec256/vec256_double.h#L278-L281
This will give different output in the edge case where `max` argument value is less than `min` and the target element value is also less than `min`.

However, I am unsure if there is a method to use in AVX for clamping efficiently and equivalently to the ternary operator. I would appreciate any advice on this.

If this inconsistency is kept unfixed, then I should also change the computation of the gradients wrt `max` and `min` arguments to take this conflict into account. The current backward computation assumes that the forward pass used ternary operator implementation.

mruberry commented:
@vsimkus You can use Google's Colab to run PyTorch on a GPU (works out of the box) and a TPU (see example Colabs at https://github.com/pytorch/xla). If your code works on CPU and GPU then the PyTorch/XLA team also often handles making it work properly for the TPU, too.

That's a great catch on the divergent behavior when max < min. Maybe clamp should make >= min (max = max(max, min) in CPP before calling a kernel?

neginraoof commented:
@vsimkus Thanks for working on this issue. Are you still working on this PR? We need this change for export of torchvision models.

vsimkus commented:
@neginraoof Hi, apologies for the delay, I had to put this on pause since I was moving countries and jobs. I will try to work on it this weekend.

The outstanding work is:
* Complete the ONNX bindings (I would appreciate someone's brain for this, as I've never had experience with ONNX before)
* Debug GPU and TPU test failures
* Fix the divergent behaviour between AVX and CPU base kernel implementations as discussed above in this PR
* Clean up tests and remove some of the debugging statements

If you really need this soon, feel free to take over this PR. Otherwise I intend to finish it although I haven't had the time recently, but hopefully should find some time now.

neginraoof commented:
Thanks! I can definitely look into ONNX issues/failures once the PR is updated. 
As far as torchvision models, so far I have suggested a couple of workarounds, but this change will simplify the code for sure.
For supporting the new clamp ops in ONNX, if it's not breaking and tests and causing failures, I can add it in a follow up PR after this is done.

vsimkus committed:
Add CUDA Aten kernels for clamp().

vsimkus commented:
@ailzhang Would you be able to look at the Pytorch/XLA build failure. It's failing on `test_clamp_raises_arg_errors` test in `test_torch.py`, which tests edge cases where none of the arguments min/max are provided. It seems like the `assertRaisesRegex` helper method doesn't handle the exceptions raised intentionally.
Otherwise, I suppose we could disable the test for XLA as an easy workaround, since it's not related to core functionality, but please let me know if there is an easy fix for testing the exception.

JackCaoG commented:
> @ailzhang Would you be able to look at the Pytorch/XLA build failure. It's failing on `test_clamp_raises_arg_errors` test in `test_torch.py`, which tests edge cases where none of the arguments min/max are provided. It seems like the `assertRaisesRegex` helper method doesn't handle the exceptions raised intentionally.
> Otherwise, I suppose we could disable the test for XLA as an easy workaround, since it's not related to core functionality, but please let me know if there is an easy fix for testing the exception.

Hmm, it seems like `test_random_from_to_xla_int32` is the one that actually failed, `test_clamp_raises_arg_errors` passed (If you scroll down a bit you will see it says ok. XLA dump the stack trace but the error is caught). I will try to take a look at why `test_random_from_to_xla_int32` seg fault. 
```
Apr 18 18:39:42   test_random_from_to_xla_int32 (__main__.TestTorchDeviceTypeXLA) ... ./test/run_tests.sh: line 42:   995 Segmentation fault      (core dumped) python3 "$CDIR/../../test/test_torch.py" "$@" -v TestTorchDeviceTypeXLA
```

dlibenzi commented:
> @ailzhang Would you be able to look at the Pytorch/XLA build failure. It's failing on `test_clamp_raises_arg_errors` test in `test_torch.py`, which tests edge cases where none of the arguments min/max are provided. It seems like the `assertRaisesRegex` helper method doesn't handle the exceptions raised intentionally.
> Otherwise, I suppose we could disable the test for XLA as an easy workaround, since it's not related to core functionality, but please let me know if there is an easy fix for testing the exception.

I think it is best to just disable the `test_clamp_raises_arg_errors` test.
Thanks for having split the error-checking tests into a separate one, so XLA can reuse the other "normal" tests!

JackCaoG commented:
I tried to patch this change on top of the pytorch master but I got bunch of compiler error related to `quantize_tensor`. Something about it couldn't find `at::quantize_tensor`. I tried to make some changes but still can't make it compiled. Do you mind try it out and let me know if it works? I used pytorch master because many of the xla failure on circleCI can be solved by a rebase on the pytorch side, and it doesn't looks like you change anything related to `normal`. Segfault is a bit concerning so I'd like to figure out if that's a real problem.

vsimkus committed:
torch.clamp with min and max as tensors

vsimkus commented:
@JackCaoG I've rebased on the latest master branch now. But it still segfaults at `test_random_from_to_xla_int32` in `test_torch.py`. Any ideas how to debug this? Let me know how I can help.
Also, why do you think that it's related to `normal`? I don't see it in the build logs.

JackCaoG commented:
@vsimkus Sorry, missed the email earlier. I will take a look at the failure today and let you know!

JackCaoG commented:
@vsimkus I patched your change on top of pytorch master. Python tests passed both individually and run with script. I think you can ignore the xla failure for now. Thanks for the patience 😄 .

vsimkus commented:
@JackCaoG Thanks for looking into it!

@mruberry Could you please review the PR again when you're available? Please let me know of any concerns you have.

In addition, below are a couple of things that this change could benefit from, but aren't included in this PR since I haven't got a complete understanding of them and am currently unable to spend lots of time on them at the moment. I suppose they can be added by someone who already is experienced with them in subsequent changes.
1. Add JIT compiler support for the `clamp_with_tensors` operations.
2. Add ONNX support (maybe @neginraoof could take a look at this once this change is landed?)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+    const Tensor& self,
+    const Tensor& min,
+    const Tensor& max) {
+  TORCH_CHECK(!self.is_complex(), "clamp is not yet implemented for complex tensors.");
```
Would you do me a favor and change these outdated "clamp is not yet implemented for complex tensors" to "clamp does not support complex inputs" ? The "yet" in the old warning (which I wrote, by the way) suggests clamp will eventually work on complex inputs, but we've since decided not to allow it.

vsimkus replied:
Sure, will do that.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+    // We clamp max tensor to make sure that it is not smaller than min
+    // This is necessary to fix the divergent behaviour in AVX and base CPU kernels
+    // https://github.com/pytorch/pytorch/pull/32587#issuecomment-591389582
+    auto adjusted_max = at::clamp_with_tensors_min(max, min);
```
I realize I'm the one who suggested this as a possible fix, but reviewing more closely now I'm worried about it because it causes divergent behavior with NumPy and may impact performance.

If we look at NumPy's implementation (https://github.com/numpy/numpy/blob/e94cec800304a6a467cf90ce4e7d3e207770b4b4/numpy/core/src/umath/clip.c.src#L78) it appears to be the same as our AVX code, and I think we should make our other implementation consistent with it, too.

vsimkus replied:
Resolved this by making all backends implement the `min(max_vec, max(min_vec, x)` formula.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+    iter.add_input(self);
+    iter.add_input(min);
+    iter.add_input(adjusted_max);
+    iter.promote_common_dtype();
```
How does promotion work in this case? If self is float but max or min are complex, does that mean I'm trying to clamp a complex tensor, which is disallowed? Should the check for complex be expanded to include the min and max tensors, too?

vsimkus replied:
Sorry, I've missing this case. Thanks for spotting it. I will add a check to include min and max tensors.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+  } else if (min.defined()) {
+    at::clamp_with_tensors_min_out(result, self, min);
+  } else {
+    AT_ERROR("At least one of 'min' or 'max' must not be None");
```
AT_ERROR is deprecated, use TORCH_CHECK(false, msg) instead.

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+  return result;
+}
+
+Tensor clamp_with_min_tensor(const Tensor& self, const Tensor& min, Scalar max) {
```
You might be able to combine some of these function definitions by optionally taking arguments like max, which would then be given as a c10::optional that you could query for whether it has a value or not.

vsimkus replied:
Hmm. I remember trying to combine them a while ago, but if I make `max` optional then the signatures in `native_functions.yaml` would conflict for function definitions: `clamp_with_tensors(Tensor self, Tensor? min=None, Tensor? max=None)` and `clamp_with_tensors.min_tensor(Tensor self, Tensor min, Scalar? max=None)`. I.e. it couldn't distinguish between a call to `clamp_with_tensors` with `min` argument only and `clamp_with_tensors.min_tensor` without `max`, and hence it complained at compile time. 

Secondly, if I were to reuse a cpp function, for example, `clamp_with_tensors_min(const Tensor& self, const Tensor& min, optional<Scalar> max)` for `native_functions.yaml` signatures `clamp_with_tensors.min_tensor(Tensor self, Tensor min, Scalar max)` and `clamp_with_tensors_min(Tensor self, Tensor min)`, how would I attach a dispatch for `clamp_with_tensors_min` to call cpp function `clamp_with_tensors_min`? I think the build would also complain regarding signature mismatch. 

I suppose the only alternative option is to change the definition of `clamp_with_tensors_min(Tensor self, Tensor min)` to `clamp_with_tensors_min(Tensor self, Tensor min, Scalar? max=None)`, and get rid of the extra `clamp_with_tensors.min_tensor` native_functions signature and the `clamp_with_min_tensor` cpp function you highlighted. However, in this case a call to `torch.clamp(min_tensor, max_scalar)` will not get automatically dispatched using the native dynamic dispatch. I would thus have to add an extra type check in `functional.py` in order to dispatch to the correct aten function.

Neither option seems perfect, so I would appreciate your thoughts.

mruberry replied:
I appreciate your thorough thinking, and confess I don't have a great idea for how to resolve the issue at this time, especially since our current clamp implementation has similar oddities. I think we can leave it for a follow-up PR.

mruberry commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -253,6 +253,33 @@ static void clamp_min_kernel(TensorIterator& iter, Scalar min_scalar) {
 }
 
+static void clamp_with_tensors_kernel(TensorIterator& iter) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(kBFloat16, iter.dtype(), "clamp_cpu", [&]() {
```
Better take out the complex dispatch

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -94,10 +94,92 @@ void erfinv_kernel_cuda(TensorIterator& iter) {
+}
+
+template<typename T>
+__host__ __device__ static inline T zabs_cuda_wrapper(thrust::complex<T> v) {
```
You don't need to handle complex. Clamping complex values based on their absolute values is probably the right way to handle them, but it doesn't follow our current clamp semantics. We'll likely have to introduce a new function or kwarg to clamp.

vsimkus replied:
Yes, I've added this before the discussions on complex behaviour happened. I could remove it, but I kept it since complex tensors are disabled in the cpp code anyway. I kept it so that the code is equivalent to the cpu implementation:
https://github.com/pytorch/pytorch/blob/e26631b33332e7409ac799470b3a4b50b4a8bdc4/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L221-L232

I am happy to remove it though too, unless there's some chance it might be useful for the further implementations.

mruberry replied:
No, let's just kill it for clarity.

mruberry commented:
@z-a-f Would you take a look at the quantization part of this PR? Or would you recommend a quantization reviewer?

mruberry commented on test/test_torch.py:
```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
         self._test_logical(device, 'logical_or', [10, 0, 1, 0], [1, 0, 0, 10], [1, 0, 1, 1])
 
-    def test_clamp(self, device):
+    def generate_clamp_baseline(self, device, min_vals=None, max_vals=None, with_nans=False):
```
Would it be easier to just compare directly with NumPy? There's a function _np_compare, in TestTorch that might be useful. Comparing with NumPy also ensures we're compatible with it (although our type promotion rules are different because we bias towards float types; that is, in Numpy float16 x int64 is float64, but in PyTorch it's float16).

vsimkus replied:
Yes, I think it would slightly more compact. I kept it this way at first to make sure the tests are the same as they were before. I can refactor them further now. Thanks for the note on type promotion, that's useful to know :)

mruberry commented on test/test_torch.py:
```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
 
-        torch.clamp(test_tens, min=min_val, out=out)
-        self.assertEqual(torch.isnan(out), torch.isnan(res1))
+    def test_clamp_min_propagates_nans(self, device):
```
You can probably roll this test into the previous test_clamp_propagates_nans by taking the product of min/max args and not passing those args to the function if they're None.

mruberry commented on test/test_torch.py:
```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
+        torch.clamp(m1, min=min_val, out=out)
+        self.assertEqual(torch.isnan(m1_base), torch.isnan(out))
 
+    def test_clamp_max_propagates_nans(self, device):
```
This test can also be combined with the previous two in a few lines, I think.

mruberry commented on test/test_torch.py:
```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
+    def test_clamp_with_tensors(self, device):
+        min_vals = torch.linspace(-1, 0, 100, device=device)  # [-1..0]
+        max_vals = torch.linspace(1, 0, 100, device=device)  # [1..0]
+        m1, m1_base = self.generate_clamp_baseline(device, min_vals, max_vals)
```
This is where you might want to follow the "expected," from NumPy, and "actual," from PyTorch, pattern found in other parts of the test suite. You'll still need a helper to your tensors vals but I think it would be simpler and (again) ensure our clamp is like NumPy's.

mruberry commented on test/test_torch.py:
```diff
@@ -6397,80 +6398,210 @@ def test_logical_and(self, device):
+        torch.clamp(m1, max=max_vals, out=out)
+        self.assertEqual(m1_base, out)
+
+    def test_clamp_with_tensors_propagates_nans(self, device):
```
I really appreciate how thorough you're being but think if there's a way to roll these tests up that only takes a few lines.

mruberry commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -538,6 +544,45 @@ Tensor clamp_backward(const Tensor & grad, const Tensor &self, const optional<Sc
+  if (!max.defined()) {
+    return grad * (self < min).type_as(grad);
+  } else {
+    // We clamp max tensor to make sure that it is not smaller than min in forward.
```
Check this with the previous note.

mruberry commented on tools/autograd/templates/Functions.cpp:
```diff
@@ -538,6 +544,45 @@ Tensor clamp_backward(const Tensor & grad, const Tensor &self, const optional<Sc
+  if (!min.defined()) {
+    return grad * (self > max).type_as(grad);
+  } else {
+    // We clamp max tensor to make sure that it is not smaller than min in forward.
```
Same as above.

mruberry commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -1335,19 +1335,64 @@ const std::vector<std::string> functions = {
-            else: #min is None and max is None
-                return grad_output, None, None
-          return torch.clamp(self, min=min, max=max), backward
+            def backward(grad_output):
```
Is this overindented?

vsimkus replied:
I don't think it is, but I will check it on an IDE. I think it was previously underindented and indented it here to fix it.

mruberry replied:
Cool. Not a big issue either way.

mruberry commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -1335,19 +1335,64 @@ const std::vector<std::string> functions = {
+                    return grad_output, None, None
+            return torch.clamp(self, min=min, max=max), backward
+
+        def clamp_with_tensors_0(self,
```
What would happen if we didn't modify this file?

vsimkus replied:
Then the `TestJitGeneratedAutograd` tests for `clamp_with_tensors` would fail, since the TorchScript and PyTorch implementation gradients must match. The tests are generated for the definitions I've added in `common_methods_invocations.py` in `method_tests`. 

As far as I know, the `symbolic_script.cpp` definitions are required for all Aten functions.

mruberry commented on torch/functional.py:
```diff
@@ -1151,5 +1153,115 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+
+    .. math::
+        y_i = \begin{cases}
+            \text{min}_i & \text{if } x_i < \text{min}_i \\
```
I think this formula will have to change based on the previous computation for the clamped value, which does allow clamping to produce values less than min.

mruberry reviewed (changes requested):
Overall this PR looks thorough and is well-written. I've requested someone more familiar with quantization take a look at that part of the PR, and I've asked a few questions and made a couple suggestions about the non-quantized part. My biggest concern is with the algorithm we use for clamping and ensuring we're consistent with NumPy and across our own implementations. Otherwise, I think the tests could probably be simplified and our clamp compared with NumPy's clip to ensure compatibility. Looking forward to hearing your thinking!

vsimkus commented:
@mruberry Thanks for the review, I appreciate your comments.

mruberry commented:
fyi @z-a-f is on vacation at the moment so he'll review or recommend a quantization reviewer later this week when he's back

z-a-f commented:
@vkuzo Can you take a look at this, to get it moving, please?

z-a-f commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,119 @@ void qclamp_kernel(
   });
 }
 
+void qclamp_with_tensors_kernel(
```
I think you need to check several things: 1) Tensor shapes being the same, 2) Quantization parameters being the same.

For number (2) I am just thinking if the tensors are quantized using different scale and zero_point, that might lead to incorrect results

vkuzo replied:
thanks for working on this!  +1, we would need to handle quantization parameters being different between qx, min_tensor, max_tensor in the kernel.

vsimkus replied:
Thanks for reviewing! I've added tests to check for different quantization parameters. I did not add tests however for different shapes as that is handled by default by the TensorIterator handler (default broadcasting rules) and is not specific to quantization, so I believe it is not necessary.

vkuzo replied:
if the quantization parameters are the same between qx, min_tensor, max_tensor then this would work.  Would be good to also handle the case when quantization params are different.  We can handle it by dequantizing, clamping in floating point, and requantizing using the quantization params of `qx`.  In pseudocode,
 
`
qx_dq = qx.dequantize();
min_tensor_dq = min_tensor.dequantize();
max_tensor_dq = max_tensor.dequantize();
... clamp in floating point ...
qy = clamped.quantize_per_tensor(...qx's quantization params...);
return qy;
`

vsimkus replied:
Thanks! I've implemented it referencing the fp implementation now + tests for that.

z-a-f commented on test/quantization/test_quantized_op.py:
```diff
@@ -456,6 +456,43 @@ def test_qclamp(self, X, min_val, max_val):
+    @given(X=hu.tensor(shapes=hu.array_shapes(1, 8, 1, 8),
+                       elements=hu.floats(-1e6, 1e6, allow_nan=False),
+                       qparams=hu.qparams()))
+    def test_qclamp_with_tensors(self, X):
```
Maybe add a tests where an error is expected to be thrown, s.a. incorrect shape, incorrect quant params, etc.

vsimkus replied:
Done (as in above comment).

vkuzo commented on test/quantization/test_quantized_op.py:
```diff
@@ -456,6 +456,43 @@ def test_qclamp(self, X, min_val, max_val):
+        }
+
+        for name, op in ops_under_test.items():
+            qY_hat = op(qX, q_min_tensor, q_max_tensor)
```
can we also test the scalar variants, and also cases where the quantization parameters of min/max tensors are not equal to quantization parameters of the input tensor?

vkuzo commented on test/quantization/test_quantized_op.py:
```diff
@@ -456,6 +456,43 @@ def test_qclamp(self, X, min_val, max_val):
+        min_tensor = np.random.uniform(-1e5, 0, size=X.shape).astype(np.float32)
+        max_tensor = np.random.uniform(0, 1e5, size=X.shape).astype(np.float32)
+
+        Y = X.copy()
```
optional: we can also reference the floating point implementation here, instead of reimplementing the calculation.  It doesn't make much difference at the current state, but might make things easier once we add handling for different quantization params.

vsimkus replied:
Now reference the floating point implementation in the tests.

vsimkus committed:
Standardized clamp to use std::min(max, std::max(min, v)) pattern in all implementations, added complex assertion for max and min arguments in clamp_with_tensors cpp dispatcher, removed complex implementation from cuda clamp kernel, added quant parameter checks

vsimkus committed:
Rolled-up clamp torch_tests, and added additional tests to test_quantized_op

vsimkus committed:
Clamp workaround for nan propagation in ROCm

vsimkus commented:
@JackCaoG Hi, would you be able to take a look into the XLA failure I'm getting on the `test_clamp_propagates_nans_xla` test, and let me know how to work around it?

The cause I know is because I have standardized clamp kernels to use `std::min(std::max(v, min), max)` pattern to match NumPy and other kernels. Whereas it previously used, depending on the kernel, either this pattern or a nested ternary operation `(v < min)? min : ((v > max)? max : v)`, which caused divergent clamp semantics between different backends when for example `max < min`. 

However, I am unsure how to fix the nan propagation (from `v`) issue on XLA. I think on XLA the nan propagation of `std::min` and `std::max` ops are different from CPU or CUDA. I had a similar issue with the ROCm backend, where it would propagate nans from the second argument in `std::min` and `std::max`, as opposed to the first argument as on CUDA/CPU. For ROCm I've added an explicit check if the `v` argument is nan, then propagate it. But I am a bit unsure where such a workaround for the XLA should go, and also whether that is the actual issue here.

Let me know your thoughts, or otherwise I would appreciate if you could point me to someone who can help!

JackCaoG commented:
@vsimkus Thanks for brining this up. I will take a look today and let you know how we want to handle it.

JackCaoG commented:
@vsimkus It seems like XLA clamp always returns min if the input is `nan`, my small repo
```
import torch                                                                                                                                                                                                                                         
import torch.nn as nn                                                                                                                                                                                                                                
import torch_xla                                                                                                                                                                                                                                     
import torch_xla.core.xla_model as xm                                                                                                                                                                                                                
from torch._six import inf, nan                                                                                                                                                                                                                      
xla_device = xm.xla_device()                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                     
t1 = torch.tensor((nan), device=xla_device)                                                                                                                                                                                                          
print(t1.clamp(-1, 1))                                                                                                                                                                                                                               
print(t1.cpu().clamp(-1, 1))                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                     
min_val = torch.tensor(-1, device=xla_device)                                                                                                                                                                                                        
max_val = torch.tensor(1, device=xla_device)                                                                                                                                                                                                         
print(t1.clamp(min_val, max_val))                                                                                                                                                                                                                    
print(t1.cpu().clamp(min_val.cpu(), max_val.cpu()))  
```
the result is 
```
tensor(-1., device='xla:0')
tensor(nan)
tensor(nan, device='xla:0')
tensor(nan)
```
The reason that second `clamp` returns `nan` is that xla has not lower the `clamp_with_tensors` nodes you added so it calls pytorch implementation. This is a known xla behavior so I think I will disable this test on the XLA side.

vsimkus commented:
@JackCaoG Thanks for looking into it! It seems to propagate the nans correctly when the min and max arguments are wrapped in pytorch-scalar (the third output in your example). I suppose I could also wrap the scalar arguments in the test - it won't change the test semantics of the test so should be ok and then it will also test the XLA behaviour. Not sure what's the preferred way: (1) keeping the test as is and having it disabled in XLA to have a reminder of the issue, or (2) since this is a known behaviour changing the test (i.e. wrapping the scalars) so that the test can be run for the XLA backend too?

dlibenzi commented:
> @JackCaoG Thanks for looking into it! It seems to propagate the nans correctly when the min and max arguments are wrapped in pytorch-scalar (the third output in your example), which is odd. I suppose I could also wrap the scalar arguments in the test - it won't change the test semantics of the test so should be ok and then it will also test the XLA behaviour. Not sure what's the preferred way: (1) keeping the test as is and having it disabled in XLA to have a reminder of the issue, or (2) since this is a known behaviour changing the test (i.e. wrapping the scalars) so that the test can be run for the XLA backend too?
> 
> EDIT: I think I misread your comment at first. Did you mean that the _third_ output returns nan because it runs the pytorch implementation and not XLA? Then yes, I think it would be best to disable the test in XLA if there's nothing else that can be done.

Getting NaN handling right in XLA is non trivial has every HW has its own, and XLA CPU uses fast-math.
It's would be nice if we could avoid having a single test, testing N different behaviors, and instead having correctly named tests which only test one behavior. Like:
```Python
def test_clamp_with_nan_input(self):
  ...
```

vsimkus commented:
@dlibenzi Thanks for the comment. The `test_clamp_propagates_nans` in fact only tests NaN propagation behaviour of clamp and clamp_with_tensors (which are ultimately the same just for different argument types) and nothing else, I think that's also what you meant? I think disabling it is fine for XLA then. 

Also, just noticed that the other test `test_clamp` that tests clamp behaviour with correct (as in non-nan) inputs is also disabled in XLA:
https://github.com/pytorch/xla/blob/0493d2d18d3e671bebbed4f1ec51365d02574aad/test/pytorch_test_base.py#L36-L37
I've refactored this test now to use NumPy as a reference implementation instead of python for-loops and hopefully it will not run slow now (as I presume it used to from the comment in XLA repo).

dlibenzi commented:
> @dlibenzi Thanks for the comment. The `test_clamp_propagates_nans` in fact only tests NaN propagation behaviour of clamp and clamp_with_tensors (which are ultimately the same just for different argument types) and nothing else, I think that's also what you meant? I think disabling it is fine for XLA then.
> 
> Also, just noticed that the other test `test_clamp` that tests clamp behaviour with correct (as in non-nan) inputs is also disabled in XLA:
> [pytorch/xla:test/pytorch_test_base.py@`0493d2d`#L36-L37](https://github.com/pytorch/xla/blob/0493d2d18d3e671bebbed4f1ec51365d02574aad/test/pytorch_test_base.py#L36-L37)
> I've refactored this test now to use NumPy as a reference implementation instead of python for-loops and hopefully it will not run slow now (as I presume it used to from the comment in XLA repo).

Yeah, every test which play with NaN is more likely not going to work than the contrary.
So is there a test we should re-enable?

vsimkus committed:
Fix manual dispatch for scalar tensors to clamp in functional.py

vsimkus commented:
@mruberry Hey, apologies for taking a long time. I've now made the changes you've requested. It was mainly two things: standardize clamping formula to use std::max(std::min(v, max), min), so it is in line with NumPy and all PyTorch kernels give same outputs, and refactored the tests to be more compact and reference NumPy for the expected behaviour. I would appreciate if you could review again!

@vkuzo  @z-a-f  Sorry, I never thanked you for the review, so thanks! I've added handling for different quantization parameters - it will now throw an error if the argument quantization parameters do not match. Also added tests to check that the error is thrown. Regarding the second point - handling different input shapes - that is already handled by the default broadcasting behaviour and errors are thrown appropriately if the argument sizes do not match, so I didn't add these specific tests to qclamp, since I don't believe that's needed.

Let me know what you think!

mruberry commented:
@z-a-f Would you review the quantization-specific portions of this PR again?

mruberry commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -607,7 +606,20 @@ template <class T,
 Vec256<T> inline clamp(const Vec256<T> &a, const Vec256<T> &min_vec, const Vec256<T> &max_vec) {
```
This looks great, but is this code path actually exercised if complex inputs to clamp are disallowed?

vsimkus replied:
I am not 100% sure. It seems clamp kernels are sometimes used in other operations, e.g. ReLU. I am not sure if there is any kernel that would require complex inputs. I can try removing this and see if there are any failing tests?

mruberry replied:
Yeah, please try removing it so we don't cause confusion.

vsimkus replied:
Ok, removed all implementations of complex clamp.

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -117,7 +118,26 @@ void clamp_kernel_cuda(TensorIterator& iter, Scalar min_value, Scalar max_value)
     auto upper = max_value.to<scalar_t>();
     gpu_kernel(iter, [=]GPU_LAMBDA(scalar_t v) -> scalar_t {
-      return (v < lower) ? lower : (v > upper ? upper : v);
+        // Propagate nan, which doesn't propagate automatically for ROCm
```
Cool. Nice comment.

mruberry commented on test/test_jit.py:
```diff
@@ -6330,11 +6330,31 @@ def test_script_clamp_min_none(x):
             return torch.clamp(x, min=2)
 
+        def test_script_clamp_with_tensors_max_none(x):
+            min = torch.full(x.shape, 2, dtype=x.dtype, device=x.device)
```
Nit: instead of using full here, what about randn?

mruberry commented on test/test_torch.py:
```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
-        max_val = 1
-        m1[1] = min_val
-        m1[2] = max_val
+    def generate_clamp_baseline(self, device, min_vals=None, max_vals=None, with_nans=False):
```
Please add a comment to describing what this function does and remove the default values for clarity but make them kwarg-only: `(self, device, * min_vals, max_vals, with_nans)`

mruberry commented on test/test_torch.py:
```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
-        self.assertEqual(out, res1)
+    def test_clamp(self, device):
+        # Create min/max argument product
+        args = product((torch.linspace(-1, 1, 100, device=device), -1, None),
```
How about generating two randn tensors instead of using linspace? Otherwise when the two tensors are selected you're only testing the narrow case when min=max, right?

vsimkus replied:
Good spot! That was a copy-paste/refactoring issue, it was meant to be `linspace(-1, 1, 100)` and `linspace(1, -1, 100)`, such that all cases are tested, I will fix it. I don't believe the randn is necessary here.

mruberry commented on test/test_torch.py:
```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
 
-        torch.clamp(m1, min=min_val, out=out)
-        self.assertEqual(out, res1)
+    def test_clamp(self, device):
```
Are there dtypes other float32 that would be interesting to test? Maybe not.

mruberry replied:
int64?

vsimkus replied:
Added int64

mruberry commented on test/test_torch.py:
```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            torch.clamp(X)
+
+    def test_clamp_propagates_nans(self, device):
```
I appreciate how thorough this is but do we actually need this test? What if `test_clamp` sets `with_nan=True`? (Unless NumPy doesn't propagate NaNs).

vsimkus replied:
I think it is necessary and it is not new - it was there before as well. The reason for test this is because different hardware/backends handle nan propagation differently. In particular, I had to make sure that ROCm backend propagated nans from the main input tensor. 

NumPy propagates nans (and this test uses NumPy as reference) but I think it is best to keep it separate from the other test because it tests a different part of functionality: `test_clamp` makes sure that the outputs are between min and max, and `test_clamp_propagates_nans` tests whether nan values in the input tensor are _always_propagated and are not masked by min/max. 

Moreover, keeping them separate allows at certain times disabling the nan test and still testing the main functionality. It has been, for example, disabled in XLA, since there is a known XLA behaviour difference in nan propagation (see above in the PR, where it was discussed with XLA team members), that is non-trivial to fix.

mruberry replied:
OK. Sounds reasonable.

mruberry commented on test/test_torch.py:
```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
-        for i in iter_indices(res2):
-            res2[i] = max(res2[i], min_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+    def test_clamp_raises_arg_errors(self, device):
```
case for clamp with complex inputs failing, too?

vsimkus replied:
Yes, it fails, throwing `clamp does not support complex inputs`, but I didn't add a test for that as I believe it would be too verbose.

mruberry commented on torch/functional.py:
```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
 
+def clamp(input, min=None, max=None, out=None):
+    r"""
+    Clamp all elements in :attr:`input` into the range `[` :attr:`min`, :attr:`max` `]` and return
```
"Clamps... and returns a tensor:"

mruberry commented on torch/functional.py:
```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+    .. math::
+        y_i = \min(\text{max}_i, \min(\text{min}_i, x_i))
+
+    The :attr:`input` and arguments :attr:`min` and attr:`max` should have matching types.
```
Is this actually true for your new functions? It looks like they support mismatched input dtypes.

vsimkus replied:
It was close, but note quite there yet. I've now made changes such that you can use mixed types and hence removed this statement.

But there is one detail - the output dtype of `clamp` and `clamp_with_tensors` can be different if the min/max types are not the same as input. For `clamp` the input dtype is preserved, and for `clamp_with_tensors` the output uses standard type promotion from all its arguments to decide the output type. I've added a statement about this in the docstring.

If you'd like to also use type promotion for `clamp` with scalar args, then I'd have to wrap both scalar arguments in `wrapped_scalar_tensor` (this seems to be the norm, for example in math operations with scalars, e.g. `add`), which I've noticed can result in significant performance decrease. And since `clamp` is a quite widely used operator (e.g. ReLU layers),  I thought it would be best to keep the original behaviour/performance in this case.

mruberry commented on torch/functional.py:
```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        y_i = \min(\text{max}_i, \min(\text{min}_i, x_i))
+
+    The :attr:`input` and arguments :attr:`min` and attr:`max` should have matching types.
+    I.e. if :attr:`input` is of type `FloatTensor` or `DoubleTensor`, then args :attr:`min`
```
The terms "FloatTensor" and DoubleTensor" are deprecated. I think this sentence can just be cut, anyway.

mruberry commented on torch/functional.py:
```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+
+    Clamps all elements in :attr:`input` to be larger or equal to :attr:`min`.
+
+    The :attr:`input` and argument :attr:`min` should have matching types.
```
See above comment.

mruberry commented on torch/functional.py:
```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+
+    Clamps all elements in :attr:`input` to be smaller or equal :attr:`max`.
+
+    The :attr:`input` and argument :attr:`max` should have matching types.
```
See above comment.

mruberry reviewed (changes requested):
Overall this looks really good and close to being done. I've requested @z-a-f review the quantized portions of the PR again, and I added a few comments, mostly on the tests and doc string.

vsimkus committed:
Allow mixed types in clamp-with-tensors, improved tests, and docstring

vsimkus committed:
Reference floating point clamp_with_tensors in qclamp

vsimkus commented:
@mruberry Thanks for the review, I've implemented the changes and added some responses to your review comments. Let me know what you think.

@vkuzo Thanks for the comment. I've changed the `qclamp_with_tensors` to reuse the fp implementation.

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,52 @@ void qclamp_kernel(
+    const Tensor& min_tensor,
+    const Tensor& max_tensor,
+    Tensor& qy) {
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "qclamp", [&]() {
```
we should be able to remove the `AT_DISPATCH_QINT_TYPES` macro from these functions.  This macro helps to codegen for different values of `SCALAR_TYPE`, but in the code we are not referencing `SCALAR_TYPE`.

vsimkus replied:
Makes sense, I wasn't sure if there wasn't anything else done by this macro. Removed.

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,52 @@ void qclamp_kernel(
+    const Tensor& min_tensor,
+    Scalar max_scalar,
+    Tensor& qy) {
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "qclamp", [&]() {
```
can remove this macro

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -631,6 +631,52 @@ void qclamp_kernel(
+    Scalar min_scalar,
+    const Tensor& max_tensor,
+    Tensor& qy) {
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "qclamp", [&]() {
```
can remove this macro

vkuzo commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -108,6 +150,38 @@ Tensor quantized_clamp(
+    const Tensor& min,
+    const Tensor& max) {
+  Tensor qy;
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "clamp", [&]() {
```
can we remove this

vkuzo commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -108,6 +150,38 @@ Tensor quantized_clamp(
+    const Tensor& min,
+    Scalar max) {
+  Tensor qy;
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "clamp", [&]() {
```
and this

vkuzo commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -108,6 +150,38 @@ Tensor quantized_clamp(
+    Scalar min,
+    const Tensor& max) {
+  Tensor qy;
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "clamp", [&]() {
```
and this

vkuzo commented on test/quantization/test_quantized_op.py:
```diff
@@ -502,32 +502,89 @@ def test_qthreshold(self, X, threshold, value):
+        X, (scale, zero_point, torch_type) = X
+        other_scale, other_zero_point, other_dtype = qparams
+
+        # Make sure that the other qparams are different
+        assume(other_dtype != torch_type
+               or other_zero_point != zero_point 
+               or other_scale != scale)
```
if we remove this check, the single test case would work for checking all the logic

vkuzo reviewed (commented):
thanks, the quantization logic looks good!  Left some comments inline, please address before landing

vsimkus committed:
Merged qclamp tests, and removed redundant dispatch macros

vsimkus commented:
@vkuzo Thanks, I've made the requested changes.

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
     func_name='lu')
 lu.__doc__ = _lu_impl.__doc__
 
+def clamp(input, min=None, max=None, out=None):
```
... max=None, *, out=None)

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        input (Tensor): the input tensor
+        min (Tensor or Number): lower-bound of the range to be clamped to
+        max (Tensor or Number): upper-bound of the range to be clamped to
+        out (Tensor, optional): the output tensor.
```
Keywords args:
{out}

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+    arguments are tensors then the output dtype uses standard type promotion rules.
+
+    Args:
+        input (Tensor): the input tensor
```
{input}

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        >>> torch.clamp(a, min=min_tensor, max=max_tensor)
+        tensor([-0.5000,  0.1734,  0.0000, -0.1000])
+
+    .. function:: clamp(input, *, min, out=None) -> Tensor
```
input, min, * out=None, right?

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+    Args:
+        input (Tensor): the input tensor
+        min (Tensor or Number): minimal value of each element in the output
+        out (Tensor, optional): the output tensor.
```
Keyword args:
out ... (also no period at end)

mruberry commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -422,13 +431,123 @@ Tensor& clamp_min_out(Tensor& result, const Tensor& self, Scalar min) {
+      .add_input(self)
+      .add_input(min)
+      .add_input(max)
+      .promote_inputs_to_common_dtype(true)
```
I would actually disable type promotion and casting to the output for simplicity. It could be enabled separately later.

mruberry replied:
(Otherwise you'll have to add another test for that behavior, but we should focus on landing this.)

mruberry commented on test/test_torch.py:
```diff
@@ -7016,80 +7017,90 @@ def test_logical_and(self, device):
-
-        # if the tensor contains nan case
-        test_tens = torch.tensor([nan], device=device)
+    def test_clamp(self, device):
```
```
@dtypes(torch.int64, torch.float32):
def test_clamp(self, device, dtype):
  ...
```

Will pass `dtype` in so you won't have to enumerate the dtypes in the function body.

vsimkus replied:
I thought a decorator like this should exist, but couldn't find it. Thanks!

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        tensor([-1.7120,  0.1734, -0.0478, -0.0922])
+        >>> torch.clamp(a, min=-0.5, max=0.5)
+        tensor([-0.5000,  0.1734, -0.0478, -0.0922])
+        >>>
```
blank line?

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+    Clamps all elements in :attr:`input` to be larger or equal to :attr:`min`.
+
+    The output dtype is same as attr:`input` if :attr:`min` argument is a scalar, otherwise if :attr:`min` is a tensor
+    then the output dtype uses standard type promotion rules.
```
Better strike these type promotion references for now (per comment above).

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+    Clamps all elements in :attr:`input` to be smaller or equal :attr:`max`.
+
+    The output dtype is same as attr:`input` if :attr:`max` argument is a scalar, otherwise if :attr:`max` is a tensor
+    then the output dtype uses standard type promotion rules.
```
More type promotion stuff

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        tensor([-0.0299, -2.3184,  2.1593, -0.8883])
+        >>> torch.clamp(a, min=0.5)
+        tensor([ 0.5000,  0.5000,  2.1593,  0.5000])
+        >>>
```
Blank line?

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        >>> torch.clamp(a, min=min_tensor)
+        tensor([-0.0299, -1.0000,  2.1593, -0.5000])
+
+    .. function:: clamp(input, *, max, out=None) -> Tensor
```
clamp(input, max, *, out=None)?

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+
+    Args:
+        input (Tensor): the input tensor
+        max (Tensor or Number): maximal value of each element in the output
```
Keyword args:
max (Tensor or Number): maximal value of each element in the output
out (Tensor, optional): the output tensor (<- no period for consistency)

mruberry commented on torch/functional.py:
```diff
@@ -1406,5 +1408,108 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
+        tensor([ 0.7753, -0.4702, -0.4599,  1.1899])
+        >>> torch.clamp(a, max=0.5)
+        tensor([ 0.5000, -0.4702, -0.4599,  0.5000])
+        >>>
```
blank line?

mruberry reviewed (changes requested):
We're in the home stretch:

- Try deleting that complex clamp impl
- Better remove type promotion for now for simplicity, it can be added later and it's not tested in the current version of the PR
- Minor doc updates

vsimkus committed:
Clamp docstring and test cleanup

vsimkus commented:
@mruberry I'm mostly done, but I'm not sure if disabling type promotion is a good way to go. First, it would throw an unfriendly error message `RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_)) INTERNAL ASSERT FAILED at "../aten/src/ATen/native/TensorIterator.cpp":168, please report a bug to PyTorch.` - of course this can be solved by adding another custom check for matching dtypes and throwing a more user friendly error. 
Second, the function would not work even for a mix of float32 and float64 dtypes, this is not good I think. So I believe that keeping the type promotion in place is better, as it seems to be standard across most ops with multiple tensor inputs I've looked at. 

I could add tests to test type promotion but I am not sure what to test. The type promotion rules are standard (i.e. what's implemented in TensorIterator) and these are already tested in `test_type_promotion.py`. I feel like adding the test here would duplicate what's done elsewhere.

Let me know what you think and how to proceed with this.

mruberry commented:
> @mruberry I'm mostly done, but I'm not sure if disabling type promotion is a good way to go. First, it would throw an unfriendly error message `RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_)) INTERNAL ASSERT FAILED at "../aten/src/ATen/native/TensorIterator.cpp":168, please report a bug to PyTorch.` - of course this can be solved by adding another custom check for matching dtypes and throwing a more user friendly error.
> Second, the function would not work even for a mix of float32 and float64 dtypes, this is not good I think. So I believe that keeping the type promotion in place is better, as it seems to be standard across most ops with multiple tensor inputs I've looked at.
> 
> I could add tests to test type promotion but I am not sure what to test. The type promotion rules are standard (i.e. what's implemented in TensorIterator) and these are already tested in `test_type_promotion.py`. I feel like adding the test here would duplicate what's done elsewhere.
> 
> Let me know what you think and how to proceed with this.

Adding a nicer error message by requiring all tensors be the same dtype sounds good for now. I appreciate your thoughtfulness on this issue but still encourage you to separate type promotion to a follow-up. If you'd really like to put it in this PR that's OK, but you'll still need to add one or more tests specifically for type promotion on clamp. These tests don't have to be crazy, and you can compare with `np.clip` which also participates in type promotion. 

Your call which way you'd like to go with it.

vsimkus committed:
Removed clamp for complex values

vsimkus committed:
Disable type promotion for clamp_with_tensors

vsimkus commented:
Thanks for the input! I've disabled type promotion for now to land this PR + added a friendlier error message. I might add the type promotion back when time allows!

mruberry reviewed (approved):
Nice work, @vsimkus. This was a lot of effort.

vsimkus commented:
Thanks for your help too!

mruberry commented:
Unfortunately our internal tests suggests this PR may be regressing performance of some popular neural networks. I'm trying to determine the cause of these regressions.

vsimkus commented:
That's likely, since clamp is used in ReLU layers as far as I remember. I didn't change the clamp with scalars much but as discussed before I have changed from ternary operator `(v < min)? min:(v> max)? max: min` to `std::min(std::max(min, v), max)` to standardize the implementations across all kernels (and NumPy), this would affect the base cpu implementation, as well as, cuda. The new implementation may be slower I believe since it can require more comparisons. I would suspect that this may be the main culprit?

mruberry commented:
> That's likely, since clamp is used in ReLU layers as far as I remember. I didn't change the clamp with scalars much but as discussed before I have changed from ternary operator `(v < min)? min:(v> max)? max: min` to `std::min(std::max(min, v), max)` to standardize the implementations across all kernels (and NumPy), this would affect the base cpu implementation, as well as, cuda. The new implementation may be slower I believe since it can require more comparisons. I would suspect that this may be the main culprit?

It's possible. If you like you can submit a PR that makes only that change for testing. It would help us diagnose the issue.

ngimel commented:
@vsimkus can you please benchmark clamp performance before and after this PR? There are benchmarking utilities available in pytorch, see [README](https://github.com/pytorch/pytorch/tree/master/torch/utils/_benchmark) and examples in that directory. The most relevant examples for you are op_benchmark.py and end_to_end.py. If UnaryOpFuzzer seems like too much, you can pick what shapes to benchmark yourself, Timer.blocked_autorange is still useful in this case to provide reliable measurements that take care of cuda synchronizations/setting num_threads.

vsimkus commented:
@ngimel Hi, I don't have a cuda device to benchmark on. I can still submit a separate PR with only the kernel changes as suggested above by @mruberry. Would that be ok?

mruberry commented:
> @ngimel Hi, I don't have a cuda device to benchmark on. I can still submit a separate PR with only the kernel changes as suggested above by @mruberry. Would that be ok?

Yep. We could use that to benchmark.
